### ISSUE TITLE
Utilise bots to actively notify on PR failure

### ISSUE THREAD

**Issue by DomT4** [Homebrew/legacy-homebrew#41103]
Discussed it a bit this morning but it'll get drowned out on the team chat and be easier to throw around as an Issue per Mike's suggestion.

Essentially have been encountering a decent chunk of builds failing, but the contributor hasn't checked back into the PR 2/3 days after they submitted it and hasn't noticed the failing red cross - Usually I grab the failure and comment and they reply back pretty quickly with either a response or a fix.

That active comment receives much quicker attention than the passive green tick/red cross yay/nay combination is an interesting possibility. We can use the bot to proactively notify people the PR has passed/failed - And hopefully reduce both the support burden on us to point out usually simple failures and the wait contributors can experience waiting for a pair of human eyes to loop over the PR. It's a potential "win-win".

The downside as Mike said on the team chat is that we have minimal control over how shouty the bot gets, and that has the potential to get *super* spammy if a PR is updated often but consistently fails or similar. We don't really want to get into a situation where we have to chase the bot around telling people it's safe to ignore xyz message because xyz.

Mike enabled it this morning, so we can see how the balance of usefulness/irritation plays out a bit.

If anyone wants to discuss it, CC @Homebrew/owners.

**Comment by xu-cheng** [comment]
Current output (see [here](https://github.com/Homebrew/homebrew/pull/41071#issuecomment-115616933)) doesn't show much useful message.

In long term, we may consider to write our own bot which can be run on Heroku. Therefore, we can custom it pretty much whatever we want:
* Send PR status comment.
* Ask user to paste `gist-log` for new issue, using keyword matching.
* Auto label e.g. new formula

etc.

**Comment by mikemcquaid** [comment]
> Current output (see here) doesn't show much useful message.

That was a single test. I've disabled/changed it now.

> In long term, we may consider to write our own bot which can be run on Heroku. Therefore, we can custom it pretty much whatever we want:
> 
> Send PR status comment.
> Ask user to paste gist-log for new issue, using keyword matching.
> Auto label e.g. new formula
> etc.

This would be an awful lot of work and we'd need to be using paid Heroku instances as the free ones won't support the load our bot is under.

**Comment by xu-cheng** [comment]
> This would be an awful lot of work

That's right. So I consider it as a long term solution.

> we'd need to be using paid Heroku instances as the free ones won't support the load our bot is under.

There are alternative services other than Heroku which offer free plan for open source organization like us. For example, [OpenShift](https://www.openshift.com/resource-grants).

**Comment by tdsmith** [comment]
I think good bot feedback would look like:

"This build failed on mavericks (_[build log](https://github.com/link/to/build/logs)_) and succeeded on el_capitan and yosemite."

My recent experimentation with snooze-button illustrated that it's comically easy to consume webhook events with AWS Lambda functions, so the problem becomes "grokking AWS" instead of "maintaining infrastructure."

**Comment by mikemcquaid** [comment]
@tdsmith Infrastructure you don't run yourself is still infrastructure you need to maintain. Again, I'd like us to avoid using services that we may outgrow and be unable to pay for before we can more or less prove that won't happen. I'd also like all new services we sign up for have more than one maintainer commit to monitoring and maintaining them to reduce the bus factor.

**Comment by DomT4** [comment]
This is something to consider at some point, but probably worth putting on hold until we decide what we're doing with CI in future.
